#2020 expression editor allow for double click to add value

veenas1 opened the pull request:
Fixes: #2020 
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

veenas1 commented on canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.scss:
```diff
@@ -184,6 +176,8 @@
 	}
 
 	.properties-vt-row-selected {
```
@matthoward366 Added style here to retain the style until the user click somewhere else. `:focus` & `:active` , is not retaining the style after clicking, and shows at the clicking moment only. 
Please let me know what looks best.

matthoward366 commented:
Overall looks good.  Checking with design on some of the interactions.

matthoward366 commented:
Keeping the cell active even after double clicking seems wrong.

![Screenshot 2024-06-25 at 1 45 25 PM](https://github.com/elyra-ai/canvas/assets/21126000/d78a7895-a00a-46a2-aae5-0f1cef928540)

veenas1 commented:
@matthoward366 Does it mean we can highlight the row(with the blue border) on single-click as well?

matthoward366 commented:
>@matthoward366 Does it mean we can highlight the row(with the blue border) on single-click as well?

single-click would highlight the row when clicking the `+` button.  The rest of the row requires double click.

veenas1 commented on canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.scss:
```diff
@@ -202,12 +212,23 @@
+				background-color: $background-active;
+			}
+		}
 	}
```
@matthoward366 I tried to add `:focus` and` :active` here. `.properties-vt-row-doubleClick` class is being detected, but the style of :focus and :active not working.

matthoward366 commented on canvas_modules/common-canvas/src/common-properties/controls/expression/expression-builder/expression-select-field-function.jsx:
```diff
@@ -56,6 +56,9 @@ export default class ExpressionSelectFieldOrFunction extends React.Component {
 		this.fields = this._makeDatasetFields(props.controller.getDatasetMetadataFields(), props.controller.getExpressionInfo().fields);
 		this.resources = get(props.controller.getExpressionInfo(), "resources", {});
 		this.state = {
+			doubleClickedFieldRow: [],
```
Why do we need to manage double click different from single click?

veenas1 replied:
@matthoward366 `properties-vt-row-doubleClick` class was added for styling the double clicked row(a blue border) which is not required for single click. 
For activating this class, I have used `doubleClickedFieldRow` from props. 
Couldn't rely on the doubleclick event alone, as single-clicking the `+` icon was supposed to act/look the same way as double-click

matthoward366 commented on canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.jsx:
```diff
@@ -439,6 +444,7 @@ class VirtualizedTable extends React.Component {
 				className={classNames(className,
 					{ "properties-vt-row-selected": selectedRow },
 					{ "properties-vt-row-disabled": rowDisabled },
+					{ "properties-vt-row-doubleClick": this.getDoubleClickedRow()?.indexOf(rowData.rowKey) > -1 },
```
Why do we need this?

veenas1 replied:
`properties-vt-row-doubleClick` class was added for styling the double clicked row(a blue border) which is not required for single click.

matthoward366 commented on canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.scss:
```diff
@@ -138,7 +138,16 @@
 			background-color: $layer-hover-01;
 		}
-
+		.properties-vt-row-doubleClick{
```
We should have row focus when a user tabs into the row but I don't think we need something special for double click.  Also, active and focus are different things.

matthoward366 commented on canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.scss:
```diff
@@ -202,12 +212,23 @@
 				border-left: 4px solid $button-primary;
 			}
 		}
+		.properties-vt-row-doubleClick{
```
This seems like duplicate code.

matthoward366 reviewed (changes requested):
I would step back and really only handle the `onRowDoubleClick` and change the pointer.  Possibly add focus for the row.  I don't think we need to worry about the `active` state.

veenas1 commented:
> I would step back and really only handle the `onRowDoubleClick` and change the pointer. Possibly add focus for the row. I don't think we need to worry about the `active` state.

ok Matt, I would revert the code added for styling the double clicked row.

matthoward366 commented:
@veenas1 let me know when the changes are done and I'll have another look.

veenas1 commented:
> @veenas1 let me know when the changes are done and I'll have another look.

@matthoward366 I have reverted all the changes added as part of styling double clicked row, the state props etc, also removed active style. The :focus is coming only on the `+` icon button

matthoward366 commented on canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.scss:
```diff
@@ -138,7 +138,10 @@
 			background-color: $layer-hover-01;
 		}
-
+		&:focus {
```
What is this for?  This might need to go under expression builder only and not the default for all tables.

matthoward366 commented on canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.scss:
```diff
@@ -189,6 +192,7 @@
 		&:hover {
 			background-color: $layer-accent-hover-01;
 		}
+		cursor: pointer;
```
When does this show up?  This might be specific to expression builder.

matthoward366 commented on canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.scss:
```diff
@@ -200,8 +204,13 @@
 		// Table rows
 		.properties-vt-row-class {
 			padding-left: 3px;
+			&:focus {
```
Same comments as above.

matthoward366 commented on canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.scss:
```diff
@@ -200,8 +204,13 @@
+			}
 		}
 		.properties-vt-row-selected {
+			cursor: pointer;
```
Same as above.

matthoward366 reviewed (changes requested):
I'm seeing the wrong mouse pointer in places outside the expression builder.  For example in the field picker the entire selected row has a pointer but it shouldn't since it's not clickable.

veenas1 commented:
@matthoward366 Updated the PR to apply changes only for expression builder's table.
